File: testCycleGanModel.py
import requests
import tempfile
import numpy as np
import os
import cv2
import zipfile
import glob

import keras.backend as K

import ants
import antspynet

K.clear_session()

dataset = 'vangogh2photo'

data_directory = os.getcwd() + "/" + dataset
if not os.path.exists(data_directory):
    zipped_file = tempfile.NamedTemporaryFile(suffix=".zip", dir=os.getcwd())
    url = "https://people.eecs.berkeley.edu/~taesung_park/CycleGAN/datasets/" + dataset + ".zip"
    r = requests.get(url)
    with open(zipped_file.name, 'wb') as f:
        f.write(r.content)

    with zipfile.ZipFile(zipped_file.name, 'r') as zip_ref:
        zip_ref.extractall(os.getcwd())

input_image_size = (128, 128, 3)

gan_model = antspynet.CycleGanModel(input_image_size=input_image_size)

print("Reading training set A files.")
trainingA_files = glob.glob(data_directory + "/trainA/*.jpg")
number_of_trainingA_files = len(trainingA_files)
X_trainA = np.zeros(shape=(number_of_trainingA_files, *input_image_size))

for i in range(number_of_trainingA_files):
   image = cv2.imread(trainingA_files[i])
   image = cv2.resize(image, input_image_size[:2])
   image = image / 127.5 - 1.0
   X_trainA[i,:,:,:] = image

print("Done.")


print("Reading training set B files.")
trainingB_files = glob.glob(data_directory + "/trainB/*.jpg")
number_of_trainingB_files = len(trainingB_files)
X_trainB = np.zeros(shape=(number_of_trainingB_files, *input_image_size))

for i in range(number_of_trainingB_files):
   image = cv2.imread(trainingB_files[i])
   image = cv2.resize(image, input_image_size[:2])
   image = image / 127.5 - 1.0
   X_trainB[i,:,:,:] = image

print("Done.")

gan_model.train(X_trainA, X_trainB, number_of_epochs=30000, sample_interval=100,
  batch_size=32, sample_file_prefix="./CycleGanSampleImages_Py/sample" )

File: testCycleGanModel.R
library( keras )
library( ANTsRNet )
library( magick )
library( ANTsR )

keras::backend()$clear_session()

dataset <- 'vangogh2photo'

dataDirectory <- paste0( getwd(), "/", dataset )
if( ! dir.exists( dataDirectory ) )
  {
  zippedFile <- tempfile( pattern = dataset, fileext = ".zip" )

  if( ! file.exists( zippedFile ) )
    {
    download.file(
      url = paste0( "https://people.eecs.berkeley.edu/~taesung_park/CycleGAN/datasets/", dataset, ".zip" ),
      destfile = zippedFile )
    unzip( zippedFile )
    }
  }

inputImageSize <- c( 128, 128, 3 )

ganModel <- CycleGanModel$new(
   inputImageSize = inputImageSize )

cat( "Reading training set A files.\n" )
trainingFilesA <- list.files( paste0( dataDirectory, '/trainA/' ),
  pattern = ".jpg", full.names = TRUE )
numberOfTrainingAFiles <- length( trainingFilesA )
X_trainA <- array( data = 0, dim = c( numberOfTrainingAFiles, inputImageSize ) )
pb <- txtProgressBar( min = 0, max = length( trainingFilesA ), style = 3 )

for( i in seq_len( numberOfTrainingAFiles ) )
  {
  image <- image_read( trainingFilesA[i] )
  image <- image_data( image_scale( image, "128x128" ) )
  image <- as.double( aperm( image, c( 3, 2, 1 ) ) )
  image <- image / 127.5 - 1.0
  X_trainA[i,,,] <- as.double( image )
  setTxtProgressBar( pb, i )
  }
cat( "\nDone.\n\n" )

cat( "Reading training set B files.\n" )
trainingFilesB <- list.files( paste0( dataDirectory, '/trainB/' ),
  pattern = ".jpg", full.names = TRUE )
numberOfTrainingBFiles <- length( trainingFilesB )
X_trainB <- array( data = 0, dim = c( numberOfTrainingBFiles, inputImageSize ) )
pb <- txtProgressBar( min = 0, max = length( trainingFilesB ), style = 3 )
for( i in seq_len( numberOfTrainingBFiles ) )
  {
  image <- image_read( trainingFilesB[i] )
  image <- image_data( image_scale( image, "128x128" ) )
  image <- as.double( aperm( image, c( 3, 2, 1 ) ) )
  image <- image / 127.5 - 1.0
  X_trainB[i,,,] <- image
  setTxtProgressBar( pb, i )
  }
cat( "\nDone.\n\n" )

ganModel$train( X_trainA, X_trainB, numberOfEpochs = 200,
  batchSize = 4, sampleInterval = 1,
  sampleFilePrefix = "./CycleGanSampleImages/sample" )

